server/shared/src/main/scala/org/http4s/server/middleware/ResponseLogger.scala

```diff
@@ -31,121 +32,165 @@ import cats.~>
+  ): ResponseLogger[F] = Impl(logHeaders, Left(logBody))
+
+  def apply[F[_]: Async](
+      logHeaders: Boolean,
+      logBodyWith: Stream[F, Byte] => Option[F[String]],
+  ): ResponseLogger[F] = Impl(logHeaders, Right(logBodyWith))
+
+  private[this] val logger = getLogger
+
+  @deprecated("Use the DSL beginning with ResponseLogger(logHeaders, logBody)", "1.0.0-M???")
   def httpApp[F[_]: Async, A](
       logHeaders: Boolean,
       logBody: Boolean,
       redactHeadersWhen: CIString => Boolean = Headers.SensitiveHeaders.contains,
       logAction: Option[String => F[Unit]] = None,
   )(httpApp: Kleisli[F, A, Response[F]]): Kleisli[F, A, Response[F]] =
-    apply(logHeaders, logBody, FunctionK.id[F], redactHeadersWhen, logAction)(httpApp)
+    apply(logHeaders, logBody)
+      .withRedactHeadersWhen(redactHeadersWhen)
+      .withLogActionOpt(logAction)(httpApp)
 
+  @deprecated("Use the DSL beginning with ResponseLogger(logHeaders, logBody)", "1.0.0-M???")
   def httpAppLogBodyText[F[_]: Async, A](
       logHeaders: Boolean,
       logBody: Stream[F, Byte] => Option[F[String]],
       redactHeadersWhen: CIString => Boolean = Headers.SensitiveHeaders.contains,
       logAction: Option[String => F[Unit]] = None,
   )(httpApp: Kleisli[F, A, Response[F]]): Kleisli[F, A, Response[F]] =
-    impl[F, F, A](logHeaders, Right(logBody), FunctionK.id[F], redactHeadersWhen, logAction)(
-      httpApp
-    )
+    apply(logHeaders, logBody)
+      .withRedactHeadersWhen(redactHeadersWhen)
+      .withLogActionOpt(logAction)(httpApp)
 
+  @deprecated("Use the DSL beginning with ResponseLogger(logHeaders, logBody)", "1.0.0-M???")
   def httpRoutes[F[_]: Async, A](
       logHeaders: Boolean,
       logBody: Boolean,
       redactHeadersWhen: CIString => Boolean = Headers.SensitiveHeaders.contains,
       logAction: Option[String => F[Unit]] = None,
   )(httpRoutes: Kleisli[OptionT[F, *], A, Response[F]]): Kleisli[OptionT[F, *], A, Response[F]] =
-    apply(logHeaders, logBody, OptionT.liftK[F], redactHeadersWhen, logAction)(httpRoutes)
+    apply(logHeaders, logBody)
+      .withRedactHeadersWhen(redactHeadersWhen)
+      .withLogActionOpt(logAction)(httpRoutes)
 
+  @deprecated("Use the DSL beginning with ResponseLogger(logHeaders, logBody)", "1.0.0-M???")
   def httpRoutesLogBodyText[F[_]: Async, A](
       logHeaders: Boolean,
       logBody: Stream[F, Byte] => Option[F[String]],
       redactHeadersWhen: CIString => Boolean = Headers.SensitiveHeaders.contains,
       logAction: Option[String => F[Unit]] = None,
   )(httpRoutes: Kleisli[OptionT[F, *], A, Response[F]]): Kleisli[OptionT[F, *], A, Response[F]] =
-    impl[OptionT[F, *], F, A](
-      logHeaders,
-      Right(logBody),
-      OptionT.liftK[F],
-      redactHeadersWhen,
-      logAction,
-    )(httpRoutes)
+    apply(logHeaders, logBody)
+      .withRedactHeadersWhen(redactHeadersWhen)
+      .withLogActionOpt(logAction)(httpRoutes)
+
+  sealed abstract class Lift[F[_], G[_]] {
```

Comment:
> Does this more likely belong to some common util module rather than `server.middleware?`

Yes possibly, but it should probably have proper laws if we expose it as something more general-purporse.

> Also, I just wonder, would it be more intuitive if we call this `HasFunctionK`? And by and large, it's a good question if we can add it to `cats` straightforwardly...

If `HasFunctionK` is just encoding that there's a canonical natural transformation from `F` to `G` does that need its own type or could it just be expressed by making certain `FunctionK` instances implicit?

Reply:
I don't think so, `FunctorK` expresses the ability to transform an algebra with a `FunctionK` instance but it doesn't provide the instance.